### ISSUE TITLE
OAuth for Gmail Access

### ISSUE THREAD

**Issue by brueyd** [pkp/pkp-lib#5365]
**Describe the problem you would like to solve**
Starting February 15, 2021, G Suite accounts will only allow access to apps using OAuth. Password-based access will no longer be supported.

**Describe the solution you'd like**
An OAuth plugin for connecting to Gmail

**Who is asking for this feature?**
System administrators that currently use Gmail SMTP and connect OJS as a "Less Secure App" to Gmail accounts for sending email notifications.

**Additional information**
Here is the complete text from the G Suite notification email...

Dear Administrator,

We’re constantly working to improve the security of your organization’s Google accounts. As part of this effort, and in consideration of the current threat landscape, we’ll be turning off access to less secure apps (LSA) — non-Google apps that can access your Google account with only a username and password, without requiring any additional verification steps. Access through only a username and password makes your account more vulnerable to hijacking attempts. Moving forward, only apps that support a more modern and secure access method called OAuth will be able to access your G Suite account.

Access to LSAs will be turned off in two stages:

June 15, 2020 - Users who try to connect to an LSA for the first time will no longer be able to do so. This includes third-party apps that allow password-only access to Google calendars, contacts, and email via protocols such as CalDAV, CardDAV and IMAP. Users who have connected to LSAs prior to this date will be able to continue using them until usage of all LSAs is turned off.

February 15, 2021 - Access to LSAs will be turned off for all G Suite accounts.

What do I need to do?
To continue using a specific app with your G Suite accounts, users in your organization must switch to a more secure type of access called OAuth. This connection method allows apps to access accounts with a digital key instead of requiring a user to reveal their username and password. We recommend that you share the user instructions (included below) with individuals in your organization to help them make the necessary changes. Alternatively, if your organization is using custom tools, you can ask the developer of the tool to update it to use OAuth. Developer instructions are also included below.

MDM configuration
If your organization uses a mobile device management (MDM) provider to configure CalDAV, CardDAV, and Exchange ActiveSync (Google Sync) profiles, these services will be phased out according to the timeline below:

June 15, 2020 - MDM push of IMAP, CalDAV, CardDAV, and Exchange ActiveSync (Google Sync) will no longer work for new users.

February 15, 2021 - MDM push of IMAP, CalDAV, CardDAV, and Exchange ActiveSync (Google Sync) will no longer work for existing users. Admins will need to push a Google Account using their MDM provider, which will re-add their Google accounts to iOS devices using OAuth.

Other less secure apps
**For any other LSA, ask the developer of the app you are using to start supporting OAuth.**
If you use other apps on iOS or MacOS that access your G Suite account information through only a password, most access issues can be resolved by removing then re-adding your account. When you add it back, make sure to select Google as the account type to automatically use OAuth.

**Comment by NateWr** [comment]
Using Gmail's SMTP is something we describe in our documentation: https://docs.pkp.sfu.ca/admin-guide/en/email#configuring-the-system-to-use-gmail-smtp

**Comment by brueyd** [comment]
Hi @NateWr. That is true, and I have been using Gmail's SMTP connection for our journal and it is working great!
  
However, **Starting February 15, 2021, G Suite accounts will only allow access to apps using OAuth. Password-based access will no longer be supported.**
  
Also, starting June 15, 2020, no new password-based "Less Secure App" connections will be allowed, which would prevent new journals from using Gmail with OJS.  
  
I am hoping that there is a way to support OAuth for Gmail before our time runs out.

**Comment by asmecher** [comment]
We use PHPMailer for email delivery; it supports Gmail's OAUTH-based APIs, as documented here:
https://github.com/PHPMailer/PHPMailer/wiki/Using-Gmail-with-XOAUTH2

The missing piece will be adding support for OJS's `config.inc.php` file to properly set PHPMailer up with the required configuration. I've hacked up a draft patch to do this -- but it's completely untested, as I don't have a domain handy to test with.

If someone could follow the above instructions with this draft patch and report the results, I'd be much obliged!

[gmail-oauth.diff.txt](https://github.com/pkp/pkp-lib/files/4100702/gmail-oauth.diff.txt)

**Comment by dbqreference** [comment]
@asmecher 

I'm testing this out on the OJS server that we have up right now. I haven't seen this get commented on and we have a journal manager who is insistent on using the internal mail features so I'm setting this to try and work on it right now. I created an issue on the Forum but no one had responded to it. I will reply with results from the login tests.

**Comment by asmecher** [comment]
@dbqreference, could you please link the forum topic here (and vice versa)? That'll help keep all the discussion related to a subject connected.

**Comment by dbqreference** [comment]
@asmecher 
Here's the forum post:
https://forum.pkp.sfu.ca/t/ojs-3-2-1-1-google-smtp-email-issues/65315?u=charlescmyers

**Comment by dbqreference** [comment]
Few key points that I'll make on this for the purpose of testing

- Allowed 'Less-secure Apps' on Google Dev account (same account that the Gmail API OAUTH token info comes from
- retrieved, scrubbed, and recreated a new Access Token for the SMTP server authentication after changing a few settings
- updated the config.inc.php file with the information asmecher created quickly

After doing these, I attempted to create and send an email through OJS. The screen currently gets stuck on the email screen without doing anything. I've updated and changed the access token, for safe measure, with no changes to the emails sending. I can provide any redacted configuration information if that would help as well as enable logs to run, as I don't have that on currently.

**Comment by asmecher** [comment]
> The screen currently gets stuck on the email screen without doing anything

This is usually an indication that there was a PHP error or exception. Check your PHP error logs?

**Comment by dbqreference** [comment]
I checked the PHP error log and the OJS error log and didn't find anything. However, the PHPMailer configuration throws the following error -
`[09-Feb-2021 00:53:22 UTC] PHP Fatal error:  Uncaught League\OAuth2\Client\Provider\Exception\IdentityProviderException: invalid_client`
I'm going to look at how I have things setup and see why it's throwing this fatal error. I double-checked my installation and configuration of PHPMailer but I possibly missed something.

**Comment by dbqreference** [comment]
I'm not sure what to be looking for here, here's the full error log information for the fatal error that gets thrown:

`[09-Feb-2021 00:53:22 UTC] PHP Fatal error:  Uncaught League\OAuth2\Client\Provider\Exception\IdentityProviderException: invalid_client in .../phpmailer/vendor/league/oauth2-google/src/Provider/Google.php:118
Stack trace:
#0 .../phpmailer/vendor/league/oauth2-client/src/Provider/AbstractProvider.php(628): League\OAuth2\Client\Provider\Google->checkResponse(Object(GuzzleHttp\Psr7\Response), Array)
#1 .../phpmailer/vendor/league/oauth2-client/src/Provider/AbstractProvider.php(537): League\OAuth2\Client\Provider\AbstractProvider->getParsedResponse(Object(GuzzleHttp\Psr7\Request))
#2 .../phpmailer/vendor/phpmailer/phpmailer/get_oauth_token.php(140): League\OAuth2\Client\Provider\AbstractProvider->getAccessToken(Object(League\OAuth2\Client\Grant\AuthorizationCode), Array)
#3 {main}
  thrown in .../phpmailer/vendor/league/oauth2-google/src/Provider/Google.php on line 118`

**Comment by asmecher** [comment]
@dbqreference, I think you're doing battle with PHPMailer and its OAuth support rather than something specific to OJS. See e.g. https://github.com/PHPMailer/PHPMailer/issues/1706 -- you might consider trying out the test script described there to see if you can get the toolset working outside of OJS.

**Comment by dbqreference** [comment]
@asmecher I tested the example document information that PHPMailer has available from their website (configured with the same information that I'm attempting to use for OJS) and it was sent with success.
`2021-02-09 22:06:32 SERVER -> CLIENT: 250 2.0.0 OK 1612908392 r190sm50589qka.54 - gsmtp
2021-02-09 22:06:32 CLIENT -> SERVER: QUIT
2021-02-09 22:06:32 SERVER -> CLIENT: 221 2.0.0 closing connection r190sm50589qka.54 - gsmtp
Message sent!`
I'm still unable to send anything from OJS though, and this is after creating an entirely new client ID from scratch based on the PHPMailer issue you linked me.
The configuration I'm using for OJS is based on the document linked above, the gmail-oauth-diff.txt. 

Here's the configuration details I have in my config.inc.php file
`
;;;;;;;;;;;;;;;;;;
; Email Settings ;
;;;;;;;;;;;;;;;;;;

[email]

; Use SMTP for sending mail instead of mail()
smtp = On

; SMTP server settings
smtp_server = smtp.gmail.com
smtp_port = 587

; Enable SMTP authentication
; Supported mechanisms: ssl, tls
smtp_auth = tls
smtp_username = "********@dbq.edu"
smtp_password = "*******"
;
; Supported smtp_authtype: RAM-MD5, LOGIN, PLAIN, XOAUTH2 (see PHPMailer AuthType)
; (Leave blank to try them in that order)
smtp_authtype = XOAUTH2

; The following are required for smtp_authtype = XOAUTH2 (e.g. GMail OAuth)
; (See https://github.com/PHPMailer/PHPMailer/wiki/Using-Gmail-with-XOAUTH2)
smtp_oauth_provider = GMail OAuth
smtp_oauth_email = *******@gmail.com
smtp_oauth_clientid = *************
smtp_oauth_clientsecret = **************
smtp_oauth_refreshtoken = ************`

Here is the configuration information that I have in the Mail.inc.php - 
`		$mailer = new \PHPMailer\PHPMailer\PHPMailer();
		$mailer->IsHTML(true);
		$mailer->Encoding = 'base64';
		if (Config::getVar('email', 'smtp')) {
			$mailer->IsSMTP();
			$mailer->Port = Config::getVar('email', 'smtp_port');
			if (($s = Config::getVar('email', 'smtp_auth')) != '') {
				$mailer->SMTPSecure = $s;
				$mailer->SMTPAuth = true;
			}
            switch (strtolower(Config::getVar('email', 'smtp_oauth_provider'))){
                case 'google': $mailer->setOAuth(new OAuth([
                        'provider' => new Google([
                            'clientId' => Config::getVar('email', 'smtp_oauth_clientid'),
                            'clientSecret' => Config::getVar('email', 'smtp_oauth_clientsecret'),
					]),
					'clientId' => Config::getVar('email', 'smtp_oauth_clientid'),
					'clientSecret' => Config::getVar('email', 'smtp_oauth_clientsecret'),
					'refreshToken' => Config::getVar('email', 'smtp_oauth_refreshtoken'),
					'userName' => Config::getVar('email', 'smtp_oauth_email'),
				]));
				case null: break; // No configured provider
				default: throw new Exception('Unknown smtp_oauth_provider in config.inc.php!');
			}
 			$mailer->AuthType = Config::getVar('email', 'smtp_authtype', '');
			$mailer->Host = Config::getVar('email', 'smtp_server');
			$mailer->Username = Config::getVar('email', 'smtp_username');
			$mailer->Password = Config::getVar('email', 'smtp_password');
			if (Config::getVar('debug', 'show_stacktrace')) {
				// debug level 3 represents client and server interaction, plus initial connection debugging
				$mailer->SMTPDebug = 3;
				$mailer->Debugoutput = 'error_log';
			}
			if (Config::getVar('email', 'smtp_suppress_cert_check')) {
				// disabling the SMTP certificate check.
				$mailer->SMTPOptions = array(
					'ssl' => array(
						'verify_peer' => false,
						'verify_peer_name' => false,
						'allow_self_signed' => true
					)
				);
			}
		}`

**Comment by asmecher** [comment]
@dbqreference, have you applied the draft patch discussed here? https://github.com/pkp/pkp-lib/issues/5365#issuecomment-577448155

**Comment by dbqreference** [comment]
@asmecher Yeah I checked the differences in both files and applied them based on that draft patch. If there was more changes to be made other than the .txt file, I haven't applied those.

**Comment by asmecher** [comment]
@dbqreference, hmm, based on the draft patch I would expect your configuration line:
```
smtp_oauth_provider = GMail OAuth
```
...to bomb out with an exception ("Unknown smtp_oauth_provider in config.inc.php!"). It expects either `google` or nothing. The fact that you aren't encountering an exception is worth investigating.

**Comment by dbqreference** [comment]
@asmecher When I originally configured it, I had "Google" set in the oauth_provider line but it didn't return an error or proceed with the request. After I changed it to "GMail OAuth", there was no difference. I can go ahead and set it back to "Google" and wait to see what else I might need to update or change based on what you might find out.

**Comment by asmecher** [comment]
@dbqreference, if a `smtp_oauth_provider` setting of `GMail OAuth` doesn't cause an exception, it means that the code in the draft patch isn't being executed at all -- there's a typo in the configuration file, or a typo in the patch I can't spot, or something else like that happening. I'd suggest investigating the lines of code changed in the draft patch to see why they aren't being executed. (It's tough to look for problems in the config file you quoted above, as the code quoting applied in your post is a little garbled -- if you can try cleaning that up I can take another look. We're looking for typos here, so it'll need to be high fidelity.)

Regards,
Alec Smecher
Public Knowledge Project Team

**Comment by dbqreference** [comment]
@asmecher I redid the entire Mail.inc.php file in order to start from template and followed the draft notes again in order to ensure I didn't look over anything. Something must have been erroneous in the text that I missed, which is great because now I'm retrieving a fatal PHP error in reference to the Mail.inc.php file itself.
`[10-Feb-2021 21:21:51 UTC] PHP Fatal error:  Uncaught Error: Class 'OAuth' not found in .../classes/mail/Mail.inc.php:484`

On that line, here's the code as it's written from the draft notes that I copied into the Mail.inc.php file - 
>			switch (strtolower(Config::getVar('email', 'smtp_oauth_provider'))) {
>				case 'google': $mailer->setOAuth(new OAuth([
>					'provider' => new Google([
>						'clientId' => Config::getVar('email', 'smtp_oauth_clientid'),
>						'clientSecret' => Config::getVar('email', 'smtp_oauth_clientsecret'),

**Comment by asmecher** [comment]
@dbqreference, great, it looks like the new code is being executed. You'll need to add this near the top of `Mail.inc.php` to resolve the `Class 'OAuth' not found` error:
```
use PHPMailer\PHPMailer\SMTP;
use League\OAuth2\Client\Provider\Google;
```

**Comment by claytondf** [comment]
@dbqreference 
Hello!
   If the solution works, could you post the procedures? I tried on the OMP and was unsuccessful.

Fail:
 SMTP connect() failed. https://github.com/PHPMailer/PHPMailer/wiki/Troubleshooting,

**Comment by dbqreference** [comment]
@asmecher @claytondf Success! I was able to successfully send the email from my OJS installation and configuration to the necessary email and have it show up.

It shows up in Junk at first but I can work with that as long as it sends the email. Attached to this comment is the file information that I configured with the `Mail.inc.php` file as well as the `config.inc.php` file originally configured. I ended up having to dig into the `GmailOAUTH.php` example file that PHPMailer included with their configuration in order to get the correct heading information that asmecher originally posted. It didn't send the email with just the `use PHPMailer\PHPMailer\SMTP;` at the top, it wanted more.

> //Import PHPMailer classes into the global namespace
use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\OAuth;
// Alias the League Google OAuth2 provider class
use League\OAuth2\Client\Provider\Google;

>//SMTP needs accurate times, and the PHP time zone MUST be set
//This should be done in your php.ini, but this is how to do it if you don't have access to that
date_default_timezone_set('Etc/UTC');

>//Load dependencies from composer
//If this causes an error, run 'composer install'
require '...phpmailer/vendor/autoload.php';

Lines 489 through 512 on `Mail.inc.php` are the draft notes information that asmecher provided, including the call at the top from lines 3 through 16, which were taken from the PHPMailer GMail OAUTH example php file. The `config.inc.php` file, also attached, only includes the EMAIL section with the configuration details that asmecher provided on the draft notes.
[Mail.txt](https://github.com/pkp/pkp-lib/files/5973115/Mail.txt)
[config.txt](https://github.com/pkp/pkp-lib/files/5973137/config.txt)


Let me know if I should provide more information regarding this. Thank you!

**Comment by asmecher** [comment]
Thanks, @dbqreference -- https://github.com/pkp/pkp-lib/commit/18c17ecd120e71a9daadaf97b3fb62bac023e4c8 should encapsulate this (for the `stable-3_3_0` branch) but anyone's confirmation is welcome!

**Comment by asmecher** [comment]
These changes will be included in OJS/OMP/OPS 3.3.0-4 when that is released, and will require config file changes in order for anything to take effect (see https://github.com/pkp/ojs/commit/43cba897d6ec5451273f56dcce71bd8548067a00 for the configuration file template changes).
